backend/Dockerfile

```diff
@@ -43,14 +43,21 @@ RUN python3 setup.py install
 WORKDIR /samples
 COPY ./samples .
 
-#We need to check that all samples have been compiled without error.
-#For find program, the -exec argument is a filter predicate just like -name. It only affects whether the file is "found", not the find's exit code.
-#One way to solve this problem is to check whether we have any python pipelines that cannot compile. Here the exit code is the number of such files:
-#RUN  bash -e -c 'exit $(find . -maxdepth 2 -name "*.py" ! -exec dsl-compile --py {} --output {}.tar.gz \; -print | wc -l)'
-#I think it's better to just use a shell loop though.
-#RUN  for pipeline in $(find . -maxdepth 2 -name '*.py' -type f); do dsl-compile --py "$pipeline" --output "$pipeline.tar.gz"; done
-#The "for" loop breaks on all whitespace, so we either need to override IFS or use the "read" command instead.
-RUN  set -e; find core tutorials -maxdepth 2 -name '*.py' -type f | while read pipeline; do python3 "$pipeline"; done
+# We need to check that all samples have been compiled without error.
+# For find program, the -exec argument is a filter predicate just like -name. It
+# only affects whether the file is "found", not the find's exit code.
+# One way to solve this problem is to check whether we have any python pipelines
+# that cannot compile. Here the exit code is the number of such files:
+# RUN  bash -e -c 'exit $(find . -maxdepth 2 -name "*.py" ! -exec dsl-compile --py {} --output {}.tar.gz \; -print | wc -l)'
+# I think it's better to just use a shell loop though.
+# RUN  for pipeline in $(find . -maxdepth 2 -name '*.py' -type f); do dsl-compile --py "$pipeline" --output "$pipeline.tar.gz"; done
+# The "for" loop breaks on all whitespace, so we either need to override IFS or
+# use the "read" command instead.
+RUN line="import kfp;kfp.components.default_base_image_or_builder='gcr.io/google-appengine/python:2020-03-31-141326";\
+    set -e; find core tutorials -maxdepth 2 -name '*.py' -type f | while read pipeline; do \
+    awk -v text="$line" '!/^#/ && !p {print text; p=1} 1' "$pipeline" && \
+    python3 "$pipeline"; \
+    done
```

Comment:
looks a magic to me, :-)

quick question, why not change the pre-built sample's definition to use the base image explictlly?

Reply:
> looks a magic to me, :-)

It just insert a provided line before the first non comment line :)

> quick question, why not change the pre-built sample's definition to use the base image explictlly?

@Ark-kun can better comment. But I guess the base image we use here should be hidden from the user-facing sample source code, because it should be treated as an implementation detail (to workaround the advanced cluster permission setup).

Reply:
> looks a magic to me, :-)

I agree it looks a bit arcane =)
The code here was a bit simpler: https://github.com/kubeflow/pipelines/pull/3422#issuecomment-611239569

Anyways I can try to change that later.

Reply:
> quick question, why not change the pre-built sample's definition to use the base image explictlly?

The sample is user-facing. The initial version of this PR was specifying the custom images, but that made the samples somewhat less simple.

With this approach we have both clean samples and reliable preloaded sample pipelines.

Reply:
> The code here was a bit simpler: #3422 (comment)

Right... actually that's the first I tried... but I hit errors. It seems like running that in Dockerfile is a bit of different.  That's why I landed on the current solution.